grayscale.py

```diff
@@ -1,336 +1,1055 @@
+
+        self.fill(Grayscale.BLACK)
+        self._copy_buffers()
+        self.init_display()
+        self.state = Grayscale_ThreadState_Starting
+        _thread.start_new_thread(self._display_thread, ())
+        while self._state[Grayscale_StateIndex_State] != Grayscale_ThreadState_Running:
+            if self._state[Grayscale_StateIndex_State] == Grayscale_ThreadState_Exception:
+                break
+            idle()
+
+
+    @property
+    def error(self):
+        return self._state[Grayscale_StateIndex_State] == Grayscale_ThreadState_Exception
+
+
+    # allow use of 'with'
+    def __enter__(self):
+        return self
+    def __exit__(self, type, value, traceback):
+        self.stop()
+
+
+    def reset(self):
+        self.res(1)
+        utime.sleep_ms(1)
+        self.res(0)
+        utime.sleep_ms(10)
+        self.res(1)
+        utime.sleep_ms(10)
+
+
+    def init_display(self):
+        self.reset()
+        self.cs(0)
+        self.dc(0)
+        # initialise as usual, except with shortest pre-charge periods and highest clock frequency
+        self.spi.write(bytearray([
+            0xae, 0x20,0x00, 0x40, 0xa1, 0xa8,63, 0xc8, 0xd3,54, 0xda,0x12, 0xd5,0xf0, 0xd9,0x11, 0xdb,0x20, 0x81,0x7f,
+            0xa4, 0xa6, 0x8d,0x14, 0xad,0x30, 0xaf]))
+        self.dc(1)
+        # clear the entire GDRAM
+        zero32 = bytearray([0] * 32)
+        for _ in range(32):
+            self.spi.write(zero32)
+        self.dc(0)
+        # set the GDRAM window
+        self.spi.write(bytearray([0x21,28,99, 0x22,0,4]))
 
-        _thread.start_new_thread(self._gsThread, ())
 
     def stop(self):
-        self.state = Grayscale.STOPPING
-        while self.state != Grayscale.STOPPED:
-            sleep_us(1000)
+        self.resume()
+        if self._state[Grayscale_StateIndex_State] == Grayscale_ThreadState_Running:
+            self._state[Grayscale_StateIndex_State] = Grayscale_ThreadState_Stopping
+            while self._state[Grayscale_StateIndex_State] != Grayscale_ThreadState_Stopped:
+                idle()
+        self.cs(1)
+        self.reset()
+        self.cs(0)
+        self.dc(0)
+        # reinitialise to the normal configuration
+        self.spi.write(bytearray([
+            0xae, 0x20,0x00, 0x40, 0xa1, 0xa8,self.height-1, 0xc8, 0xd3,0, 0xda,0x12, 0xd5,0x80,
+            0xd9,0xf1, 0xdb,0x20, 0x81,0x7f,
+            0xa4, 0xa6, 0x8d,0x14, 0xad,0x30, 0xaf,
+            0x21,28,99, 0x22,0,4]))
+        self.cs(1)
 
-    def saveConfig(self):
-        ConfigFile.save(self.config)
 
-    ### Be a proxy to GraphicsClass
+    def pause(self):
+        if self.error:
+            return False
+        if self._state[Grayscale_StateIndex_State] == Grayscale_ThreadState_Paused:
+            return False
+        self._state[Grayscale_StateIndex_State] = Grayscale_ThreadState_Pausing
+        while self._state[Grayscale_StateIndex_State] != Grayscale_ThreadState_Paused:
+            idle()
+        return True
```

Comment:
Is there any particular reason why pause has a return value, and the others don't?

Reply:
Yeah, the intent here is that if you call `pause()` you keep the return value, and at the end of the paused section you then only call `resume()` if `pause()` returned True.
This could also be implemented by keeping a 'pause count' instance variable, but I thought making it the caller's responsibility would be sufficient.

FWIW the reason for having the thread pause functionality was to avoid a crash when performing file I/O. This doesn't actually seem to be necessary, so the pause stuff can probably be removed. I can't really think of any other use-case for it. Can you?

Reply:
Understood! Not immediately. Stopping the thread entirely and restarting it would effectively achieve the same.